### PR TITLE
Properly restore the "orphaned" tooltips

### LINKED ISSUE #3696: Investigate the "orphaned" tooltips
Got a HUGE amount of tool tips after several patch changes.

137 tool tip groups, many having 10+ tool tips.
image

The worst:
image

It is obvious that some code which adds a tooltip fails to remove it.

### PR DESCRIPTION
**Pull request by 84634E1A607A**
Some tooltips are not removed correctly when related nodes are disposed.

**Related Issues**
Fixes #3696

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by 84634E1A607A** [comment]
This further reduced the lag when changing patch, now that tooltips are correctly restored, and we don't need to request for new tooltip nodes each and every patch change.

File: src/gui_common/charts/line/LineChart.cs [hunk 1]
```diff
@@ -385,6 +395,9 @@ public void ClearDataSets()
 
             // Initialize line
             var dataLine = new DataLine((LineChartData)data.Value, data.Key == defaultDataSet, isChild ? 2 : 1);
+            dataLine.Connect("tree_exited", this, nameof(RemoveDataLineTooltipOnDeletion), new Array(dataLine));
+
+            dataLineTooltips.Add(dataLine, new List<ICustomToolTip>());
```
Comment (Kasterisk) [inline src/gui_common/charts/line/LineChart.cs:400]: For consistency and extra safety, add check for existing keys.
https://github.com/Revolutionary-Games/Thrive/blob/0f71e7936bbc2adef1bc1480c7418231202cb2a8/src/gui_common/charts/line/LineChart.cs#L391-L392

File: src/gui_common/charts/line/LineChart.cs [hunk 2]
```diff
@@ -385,6 +395,9 @@ public void ClearDataSets()
 
             // Initialize line
             var dataLine = new DataLine((LineChartData)data.Value, data.Key == defaultDataSet, isChild ? 2 : 1);
+            dataLine.Connect("tree_exited", this, nameof(RemoveDataLineTooltipOnDeletion), new Array(dataLine));
```
Comment (Kasterisk) [inline src/gui_common/charts/line/LineChart.cs:398]: I think this should be more general, that is to connect `LineChart`'s own `tree_exited` signal to a new method called `ClearTooltips` to make sure everything else gets destroyed. The tooltip clearing logic in `ClearChart` needs to be moved there and the former just simply call `ClearTooltips`.

File: src/gui_common/charts/line/LineChart.cs [hunk 3]
```diff
@@ -620,6 +633,14 @@ public void OverrideDataPointToolTipDescription(string dataset, DataPoint datapo
         }
     }
 
+    private void RemoveDataLineTooltipOnDeletion(DataLine line)
+    {
+        foreach (var tooltip in dataLineTooltips[line])
+        {
+            ToolTipManager.Instance.RemoveToolTip(tooltip.ToolTipNode.Name, LINE_CHART_GROUP_BASE_NAME);
+        }
+    }
```
Comment (Kasterisk) [inline src/gui_common/charts/line/LineChart.cs:642]: It's rarely that we need to remove a specific dataset and thus each line's tooltip separately so an all in one solution seems fitting.

**Review by Kasterisk — CHANGES_REQUESTED** [review]

**Comment by Kasterisk** [comment]
Reading through the code again, it seems like my past self made each line segments have their own tooltips in the _same_ dataset. That's rather inefficient and probably needs optimization. Commented here for future records.

File: src/gui_common/charts/line/LineChart.cs (hunk 2 above)
Comment (84634E1A607A) [reply src/gui_common/charts/line/LineChart.cs:398]: LineChart delete lines when data is updated. (And that was the main cause of orphaned tooltips).

Comment (Kasterisk) [reply src/gui_common/charts/line/LineChart.cs:398]: Yes that's because `ClearChart` is called which among other things also clears the tooltips.

**Review by Kasterisk — APPROVED** [review]
Well, good enough I guess.

**Pull request merged**